moveit_ros/perception/handeye_calibration_target/src/handeye_aruco_target.cpp

```diff
@@ -0,0 +1,301 @@
+                                        << "border_bits" << std::to_string(border_bits) << "\n"
+                                        << "dictionary_id " << dictionary_id << "\n");
+  if (markers_x > 0 && markers_y > 0 && marker_size > 0 && separation > 0 && border_bits > 0 && !dictionary_id.empty())
+  {
+    std::lock_guard<std::mutex> lck(aruco_mtx_);
+    markers_x_ = markers_x;
+    markers_y_ = markers_y;
+    marker_size_ = marker_size;
+    separation_ = separation;
+    border_bits_ = border_bits;
+
+    const auto& it = marker_dictionaries_.find(dictionary_id);
+    if (it != marker_dictionaries_.end())
+      dictionary_id_ = it->second;
+    else
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "Invalid marker dictionary name.");
+      return false;
+    }
+  }
+  else
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Invalid target intrinsic params.");
+    return false;
+  }
+
+  return true;
+}
+
+bool HandEyeArucoTarget::setTargetDimension(const double& marker_measured_size,
+                                            const double& marker_measured_separation)
+{
+  if (marker_measured_size != marker_size_real_ || marker_measured_separation != marker_separation_real_)
+    ROS_DEBUG_STREAM_NAMED(LOGNAME, "Set target real dimensions: "
+                                        << "\n"
+                                        << "marker_measured_size " << std::to_string(marker_measured_size) << "\n"
+                                        << "marker_measured_separation " << std::to_string(marker_measured_separation)
+                                        << "\n");
+
+  if (marker_measured_size > 0 && marker_measured_separation > 0)
+  {
+    std::lock_guard<std::mutex> lck(aruco_mtx_);
+    marker_size_real_ = marker_measured_size;
+    marker_separation_real_ = marker_measured_separation;
+    return true;
+  }
+  else
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Invalid target real dimensions.");
+    return false;
+  }
+}
+
+bool HandEyeArucoTarget::createTargetImage(cv::Mat& image)
+{
+  cv::Size image_size;
+  image_size.width = markers_x_ * (marker_size_ + separation_) - separation_ + 2 * separation_;
+  image_size.height = markers_y_ * (marker_size_ + separation_) - separation_ + 2 * separation_;
+
+  try
+  {
+    // Create target
+    cv::Ptr<cv::aruco::Dictionary> dictionary = cv::aruco::getPredefinedDictionary(dictionary_id_);
+    cv::Ptr<cv::aruco::GridBoard> board =
+        cv::aruco::GridBoard::create(markers_x_, markers_y_, float(marker_size_), float(separation_), dictionary);
+
+    // Create target image
+    board->draw(image_size, image, separation_, border_bits_);
+  }
+  catch (cv::Exception& e)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Aruco target image creation exception: " << e.what());
+    return false;
+  }
+
+  return true;
+}
+
+bool HandEyeArucoTarget::detectTargetPose(cv::Mat& image)
+{
+  std::lock_guard<std::mutex> lck(base_mtx_);
```

Comment:
Since the target or the camera intrinsic parameters could be changed by the user during the detection running. I would like to use the locks to make sure they are not changed in the middle of a detection.